### PR TITLE
Forbid sending native histograms between components in JSON format

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

In order to support query sharding of queries that include native histograms, as well as use of the new protobuf format between query-frontends and rulers, we have to ensure that the implicit native histogram format is used between queriers and query-frontends.

Prometheus' JSON format uses the explicit format, so we have to block sending native histograms if the JSON format is in use.

#### Which issue(s) this PR fixes or relates to

https://github.com/grafana/mimir/issues/4104

#### Checklist

- [x] Tests updated
- [n/a] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by charleskorn** [comment]
> `CanEncode` seems to be doing the right thing, although I don't know how it all fits in.

`CanEncode` is used by [`API.negotiateCodec()`](https://github.com/grafana/mimir/blob/c261e50438c55cf2f3913bc5189f9a8956b03f92/vendor/github.com/prometheus/prometheus/web/api/v1/api.go#L1606) to decide whether or not to use the `Codec` to encode the response.

**Review by krajorama — COMMENTED** [review]
I don't understand why we'd need this.

We will only use the protobuf between query-frontend and querier, because we must support native histograms efficiently. This covers codec when query-frontend decodes and querier encodes.

On the query-frontend we must be able to return results in JSON (and optionally protobuf), because that's what Grafana and most clients will use, so that should allow for both JSON and protobuf where JSON contains the explicit buckets of histograms. This covers query-frontend encoding and something else decoding.

File: pkg/api/handlers.go [hunk 1]
```diff
@@ -261,6 +261,8 @@ func NewQuerierHandler(
 		nil,
 	)
 
+	api.ClearCodecs()
+	api.InstallCodec(mimirJSONCodec{})
 	api.InstallCodec(protobufCodec{})
```
Comment (krajorama) [inline pkg/api/handlers.go:266]:
```suggestion
	api.ClearCodecs()
	api.InstallCodec(protobufCodec{})
```

Wouldn't this achieve the same? See my overall comment.

**Comment by charleskorn** [comment]
> I don't understand why we'd need this.
> 
> We will only use the protobuf between query-frontend and querier, because we must support native histograms efficiently. This covers codec when query-frontend decodes and querier encodes.
> 
> On the query-frontend we must be able to return results in JSON (and optionally protobuf), because that's what Grafana and most clients will use, so that should allow for both JSON and protobuf where JSON contains the explicit buckets of histograms. This covers query-frontend encoding and something else decoding.

This PR only affects the behaviour of queriers. Query-frontends will continue to send JSON with histograms in explicit format. 

(Queriers use the encoding methods in `pkg/api`, whereas query-frontends use the encoding and decoding methods in `pkg/frontend`. We don't use the same code to encode and decode everywhere because different components use different internal representations of query results. For example, queriers use `promql.Vector`, `promql.Matrix` etc., while the query-frontend uses a single `PrometheusData` type.)

File: pkg/api/handlers.go (hunk 1 above)
Comment (charleskorn) [reply pkg/api/handlers.go:266]: We could do this in the future, but for now, we need to continue supporting JSON responses in queriers for two scenarios:
* For clusters where query-frontends do not have the protobuf format enabled
* To support easy upgrades between versions - if we remove support for JSON responses in queriers, then we force users to upgrade their query-frontends and enable the protobuf format first, then upgrade queriers separately. If we support both formats, then both components can be rolled out simultaneously, and queriers that support protobuf will only send it to query-frontends that report they support it (via the `Accept` header).

Comment (krajorama) [reply pkg/api/handlers.go:266]: My concern is that we're adding a linear algorithm, what is the impact on the performance and resource utilization ? 

I think another way to solve this is to introduce the protobuf encoder, but do not enable native histogram ingestion for anybody, then in the next upgrade turn off the JSON encoder. After the second upgrade you can enable histograms.

Or even better, we could have the MimirJSON encoder be able to encode histograms in the internal format and done.

Comment (charleskorn) [reply pkg/api/handlers.go:266]: Yep, I agree, I think I've been overzealous here :) We can leave the JSON encoder as it is, and the worst case scenario is that either histograms results disappear (if the query-frontend is too old) or the query-frontend rejects the response containing explicit histograms [here](https://github.com/grafana/mimir/blob/4b43206a721088cdcbdeb1fc943d5797f2b13d45/pkg/frontend/querymiddleware/model_extra.go#L373) (if the query-frontend has been upgraded).

**Comment by charleskorn** [comment]
Not needed - see discussion here: https://github.com/grafana/mimir/pull/4317#discussion_r1125770492.

**Pull request closed without merging**